trin-core/src/types/receipts.rs

```diff
@@ -0,0 +1,523 @@
+use std::ops::{Deref, DerefMut};
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use bytes::Bytes;
+use eth_trie::{EthTrie, MemoryDB, Trie};
+use ethereum_types::{Address, Bloom, BloomInput, H256, U256};
+use rlp::{self, DecoderError, Rlp, RlpStream};
+use rlp_derive::{RlpDecodable, RlpEncodable};
+
+/// Represents the `Receipts` datatype used by the chain history wire protocol
+pub struct Receipts {
+    pub receipt_list: Vec<TypedReceipt>,
+}
+
+impl Receipts {
+    pub fn root(&self) -> anyhow::Result<H256> {
+        let memdb = Arc::new(MemoryDB::new(true));
+        let mut trie = EthTrie::new(Arc::clone(&memdb));
+
+        // Insert receipts into receipts tree
+        for (index, receipt) in self.receipt_list.iter().enumerate() {
+            let path = rlp::encode(&index).freeze().to_vec();
+            let encoded_receipt = receipt.encode();
+            match trie.insert(&path, &encoded_receipt) {
+                Ok(_) => (),
+                Err(msg) => return Err(anyhow!("Error calculating receipts root: {msg:?}")),
+            }
+        }
+
+        match trie.root_hash() {
+            Ok(val) => Ok(val),
+            Err(msg) => Err(anyhow!("Error calculating receipts root: {msg:?}")),
+        }
+    }
+}
+
+// Based on https://github.com/openethereum/openethereum/blob/main/crates/ethcore/types/src/receipt.rs
+
+/// A record of execution for a `LOG` operation.
+#[derive(Default, Debug, Clone, PartialEq, Eq, RlpEncodable, RlpDecodable)]
+pub struct LogEntry {
+    /// The address of the contract executing at the point of the `LOG` operation.
+    pub address: Address,
+    /// The topics associated with the `LOG` operation.
+    pub topics: Vec<H256>,
+    /// The data associated with the `LOG` operation.
+    pub data: Bytes,
+}
+
+impl LogEntry {
+    /// Calculates the bloom of this log entry.
+    pub fn bloom(&self) -> Bloom {
+        self.topics.iter().fold(
+            Bloom::from(BloomInput::Raw(self.address.as_bytes())),
+            |mut b, t| {
+                b.accrue(BloomInput::Raw(t.as_bytes()));
+                b
+            },
+        )
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TransactionOutcome {
+    /// Status and state root are unknown under EIP-98 rules.
+    Unknown,
+    /// State root is known. Pre EIP-98 and EIP-658 rules.
```

Comment:
As far as I can tell, EIP-98 was never deployed. We went straight to EIP-658. Would means this seems sufficient:
```suggestion
    /// State root is known, before EIP-658 is enabled.
```